internal/access/grant.go

```diff
@@ -31,18 +41,103 @@ func ListGrants(c *gin.Context, opts data.ListGrantsOptions) ([]models.Grant, er
 		subjectID, _ := subject.ID() // zero value will never match a user
 		switch {
 		case rCtx.Authenticated.User == nil:
-			return nil, err
+			return ListGrantsResponse{}, err
 		case subject.IsIdentity() && rCtx.Authenticated.User.ID == subjectID:
 			// authorized because the request is for their own grants
 		case subject.IsGroup() && userInGroup(rCtx.DBTxn, rCtx.Authenticated.User.ID, subjectID):
 			// authorized because the request is for grants of a group they belong to
 		default:
-			return nil, err
+			return ListGrantsResponse{}, err
 		}
 	} else if err != nil {
-		return nil, err
+		return ListGrantsResponse{}, err
+	}
+
+	if lastUpdateIndex == 0 {
+		result, err := data.ListGrants(rCtx.DBTxn, opts)
+		return ListGrantsResponse{Grants: result}, err
+	}
+
+	// Close the request scoped txn to avoid long-running transactions.
+	if err := rCtx.DBTxn.Rollback(); err != nil {
```

Comment:
Should we commit this transaction rather than rolling it back? If we modify the database somewhere else it seems potentially unexpected that it gets rolled back here when there is no error.

Reply:
This is in `ListGrants` so it should always be a read-only db connection. `get` in `routes.go` is used to create this route, and it sets `txnOptions` to `ReadOnly: true`. Any writes would fail before this point I think.  

We shouldn't be writing anything in `GET` requests, except for the "middleware" stuff at the beginning of a request, which we do in a separate transaction.

I'm not sure if there is a difference between `Commit` nothing , or `Rollback` nothing, but I guess rollback does communicate to the reader that a bit better.